Mathlib/CategoryTheory/FiberedCategory/HasFibers.lean

```diff
@@ -0,0 +1,248 @@
+equivalent to the fibers `Fiber p S`. One would then like to use these categories `Fib S` directly,
+instead of working through this equivalence of categories. By developing an API for the `HasFibers`
+typeclass, this will be possible.
+
+Here is an example of when this typeclass is useful. Suppose we have a presheaf of types
+`F : 𝒮ᵒᵖ ⥤ Type _`. The associated fibered category then has objects `(S, a)` where `S : 𝒮` and `a`
+is an element of `F(S)`. The fiber category `Fiber p S` is then equivalent to the discrete category
+`Fib S` with objects `a` in `F(S)`. In this case, the `HasFibers` instance is given by the
+categories `F(S)` and the functor `ι` sends `a : F(S)` to `(S, a)` in the fibered category. See
+`Presheaf.lean` for more details.
+
+## Main API
+The following API is developed so that the fibers from a `HasFibers` instance can be used
+analogously to the standard fibers.
+
+- `mapPreimage φ` is a lift of a morphism `φ : (ι S).obj a ⟶ (ι S).obj b` in `𝒳`, which lies over
+`𝟙 S`, to a morphism in the fiber over `S`.
+- `objPreimage` gives an object in the fiber over `S` which is isomorphic to a given `a : 𝒳` that
+satisfies `p(a) = S`. The isomorphism is given by `objObjPreimageIso`.
+- `HasFibers.pullbackObj` is a version of `IsPreFibered.pullbackObj` which ensures that the object
+lies in a given fiber. The corresponding cartesian morphism is given by `HasFibers.pullbackMap`.
+- `HasFibers.inducedMap` is a version of `IsCartesian.inducedMap` which gives the corresponding
+morphism in the fiber category.
+- `fiber_factorization` is the statement that any morphism in `𝒳` can be factored as a morphism in
+some fiber followed by a pullback.
+
+-/
+
+universe v₁ u₁ v₂ u₂ v₃ u₃
+
+open CategoryTheory Functor Category IsCartesian IsHomLift Fiber
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category.{v₁} 𝒮] [Category.{v₂} 𝒳]
+
+/-- HasFibers is an extrinsic notion of fibers on a functor `p : 𝒳 ⥤ 𝒮`. It is given by a
+collection of categories `Fib S` for every `S : 𝒮` (the fiber categories), each equiped with a
+functors `ι : Fib S ⥤ 𝒳` which map constantly to `S` on the base such that the induced functor
+`Fib S ⥤ Fiber p S` is an equivalence. -/
+@[nolint checkUnivs]
+class HasFibers (p : 𝒳 ⥤ 𝒮) where
+  /-- The type of objects of the category `Fib S` for each `S`. -/
+  Fib (S : 𝒮) : Type u₃
+  /-- `Fib S` is a category. -/
+  isCategory (S : 𝒮) : Category.{v₃} (Fib S) := by infer_instance
+  /-- The functor `ι : Fib S ⥤ 𝒳`. -/
+  ι (S : 𝒮) : (Fib S) ⥤ 𝒳
+  /-- The composition with the functor `p` is *equal* to the constant functor mapping to `S`. -/
+  comp_const (S : 𝒮) : (ι S) ⋙ p = (const (Fib S)).obj S
+  /-- The induced functor from `Fib S` to the fiber of `𝒳 ⥤ 𝒮` over `S` is an equivalence. -/
+  equiv (S : 𝒮) : Functor.IsEquivalence (inducedFunctor (comp_const S))
+
+namespace HasFibers
+
+/-- The `HasFibers` on `p : 𝒳 ⥤ 𝒮` given by the fibers of `p` -/
+@[default_instance]
+instance canonical (p : 𝒳 ⥤ 𝒮) : HasFibers p where
```

Comment:
I think this should not be an instance, only a `def`.

Reply:
Then will it still work as a "default instance"?

Reply:
I am not sure exactly what `default_instance` do. I think we should remove it.

Reply:
My opinion is that I would not like to have default instances flying around... If we need a `HasFibers` instance in order to prove a statement (which does not already involve a `HasFibers`), then we may invoke this `canonical` in order to make the argument work. Otherwise, `[HasFibers p]` should be a variable.